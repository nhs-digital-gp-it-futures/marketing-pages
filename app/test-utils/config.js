export const apiLocalhost = 'http://localhost:8080';

export const apiPath = '/api/v1/Solutions/S100000-001';

export const clientLocalhost = 'http://localhost:1234';
