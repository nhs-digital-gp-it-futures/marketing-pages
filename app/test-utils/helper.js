export const extractInnerText = async (element) => {
  const elementInnerText = await element.innerText;
  return elementInnerText.trim();
};
