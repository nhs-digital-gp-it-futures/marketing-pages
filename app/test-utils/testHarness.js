/* eslint-disable import/no-extraneous-dependencies */
import express from 'express';
import request from 'supertest';
import cheerio from 'cheerio';
import nunjucks from 'nunjucks';
import { App } from '../../app';

const testFunction = ({ setup, done }) => {
  const app = new App().createApp();
  const router = express.Router();

  const macroWrapper = setup.component ? `{% from '${setup.component.path}' import ${setup.component.name} %}
                      {{ ${setup.component.name}(params) }}` : '';

  return {
    request: (context, callback) => {
      const dummyRouter = router.get('/', (req, res) => {
        if (setup.template) {
          res.render(setup.template.path, context);
        } else {
          const viewToTest = nunjucks.renderString(macroWrapper, context);
          res.send(viewToTest);
        }
      });
      app.use(dummyRouter);

      request(app).get('/').then((response) => {
        callback(cheerio.load(response.text));
        done();
      });
    },
  };
};

export const createTestHarness = (setup, callback) => (done) => {
  callback(testFunction({
    setup,
    done,
  }));
};
