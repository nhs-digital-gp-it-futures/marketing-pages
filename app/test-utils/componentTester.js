import nunjucks from 'nunjucks';
import { createTestHarness } from 'buying-catalogue-library';
import { App } from '../app';
import config from '../config';

export const componentTester = (setup, callback) => (done) => {
  const app = new App().createApp();

  callback(createTestHarness({
    app,
    templateEngine: nunjucks,
    config,
    setup,
    done,
  }));
};
