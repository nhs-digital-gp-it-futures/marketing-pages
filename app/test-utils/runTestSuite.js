import nock from 'nock';
import { getSectionManifest } from '../manifestProvider';
import { runCommonComponentsTests } from './testFunctions/commonComponentsTests';
import { runQuestionTests } from './testFunctions/questionTests';
import { apiLocalhost, apiPath, clientLocalhost } from './config';

export const runTestSuite = async ({
  data,
  sectionId,
  dashboardId,
  errorPostBodyData,
  userContextType = 'supplier',
}) => {
  const clientUrl = dashboardId ? `${clientLocalhost}/${userContextType}/solution/S100000-001/dashboard/${dashboardId}/section/${sectionId}`
    : `${clientLocalhost}/${userContextType}/solution/S100000-001/section/${sectionId}`;

  const mocks = async (responseStatus, responseBody) => {
    await nock(apiLocalhost)
      .get(`${apiPath}/sections/${sectionId}`)
      .reply(responseStatus, responseBody);
  };

  const pageSetup = async ({
    t,
    responseStatus = 200,
    responseBody = {},
  }) => {
    await mocks(responseStatus, responseBody);
    await t.navigateTo(clientUrl);
  };

  const sectionManifest = getSectionManifest({ sectionId, dashboardId, userContextType });

  fixture(`Show ${sectionManifest.title} page`);

  await Promise.all([
    runCommonComponentsTests({
      pageSetup,
      sectionManifest,
      sectionId,
      data,
      dashboardId,
      userContextType,
    }),
    runQuestionTests({
      pageSetup,
      sectionManifest,
      data,
      sectionId,
      dashboardId,
      errorPostBodyData,
    }),
  ]);
};
