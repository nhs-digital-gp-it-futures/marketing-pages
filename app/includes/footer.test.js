import { componentTester } from '../test-utils/componentTester';

const setup = {
  template: {
    path: 'includes/footer.njk',
  },
};

const footerLinks = [
  {
    label: 'Buyer\'s Guide',
    URL: '/guide',
  },
  {
    label: 'NHS Digital Helpdesk',
    URL: '/guide#contact-us',
  },
  {
    label: 'NHS Digital',
    URL: 'https://digital.nhs.uk/',
  },
  {
    label: 'About GP IT Futures',
    URL: 'https://digital.nhs.uk/services/future-gp-it-systems-and-services',
  },
  {
    label: 'Capabilities & Standards Model',
    URL: 'https://gpitbjss.atlassian.net/wiki/spaces/GPITF/overview',
  },
];

describe('footer', () => {
  it('should render the footer panel', componentTester(setup, (harness) => {
    const context = { footerLinks };

    harness.request(context, ($) => {
      const footer = $('[data-test-id="footer"]');
      const footerComponent = footer.find('[data-test-id="footer-component"]');

      expect(footer.length).toEqual(1);
      expect(footerComponent.length).toEqual(1);

      footerLinks.map((link, i) => {
        expect(footerComponent.find(`li:nth-child(${i + 1})`).text().trim()).toEqual(link.label);
      });
    });
  }));

  it('should render the footer legal panel', componentTester(setup, (harness) => {
    const context = { showLegalPane: false };

    harness.request(context, ($) => {
      const footer = $('[data-test-id="footer"]');
      const legalPanel = footer.find('[data-test-id="legal-panel"]');
      expect(footer.length).toEqual(1);
      expect(legalPanel.length).toEqual(1);
      expect(legalPanel.find('span:nth-child(1)').text().trim()).toEqual('Legal');
      expect(legalPanel.find('span:nth-child(2)').text().trim()).toEqual('Privacy and Cookies');
      expect(legalPanel.find('span:nth-child(2) > a').attr('href')).toEqual('https://digital.nhs.uk/about-nhs-digital/privacy-and-cookies');
    });
  }));
});
