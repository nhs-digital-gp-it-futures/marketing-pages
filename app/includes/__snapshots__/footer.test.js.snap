// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`footer should render the footer 1`] = `
"<footer role=\\"contentinfo\\" data-test-id=\\"footer\\" class=\\"bc-bg--grey-4\\">
    <div class=\\"nhsuk-footer\\" id=\\"nhsuk-footer\\">
      <div class=\\"nhsuk-width-container\\">
        <h2 class=\\"nhsuk-u-visually-hidden\\">Support links</h2>
        <ul class=\\"nhsuk-footer__list\\">
          <li class=\\"nhsuk-footer__list-item\\"><a class=\\"nhsuk-footer__list-item-link\\" href=\\"/guide\\">Buyer's Guide</a></li>
          <li class=\\"nhsuk-footer__list-item\\"><a class=\\"nhsuk-footer__list-item-link\\" href=\\"/guide#contact-us\\">NHS Digital Helpdesk</a></li>
          <li class=\\"nhsuk-footer__list-item\\"><a class=\\"nhsuk-footer__list-item-link\\" href=\\"https://digital.nhs.uk/\\">NHS Digital</a></li>
          <li class=\\"nhsuk-footer__list-item\\"><a class=\\"nhsuk-footer__list-item-link\\" href=\\"https://digital.nhs.uk/services/future-gp-it-systems-and-services\\">About GP IT Futures</a></li>
          <li class=\\"nhsuk-footer__list-item\\"><a class=\\"nhsuk-footer__list-item-link\\" href=\\"https://gpitbjss.atlassian.net/wiki/spaces/DCSH/overview\\">Capabilities and Standards models</a></li>
        </ul>
        
        <p class=\\"nhsuk-footer__copyright\\">Â© Crown copyright</p>
      </div>
    </div>

    
      <ul data-test-id=\\"legal-panel\\" class=\\"bc-lh-r bc-body-r bc-list-none nhsuk-u-padding-0 nhsuk-u-padding-bottom-3 nhsuk-width-container\\">
        <li class=\\"bc-u-display-inline nhsuk-u-margin-0 nhsuk-u-margin-right-2\\">Legal</li>
        <li class=\\"bc-u-display-inline nhsuk-u-margin-0 nhsuk-u-margin-right-2\\"><a href=\\"/privacy-policy\\">Privacy policy and cookies</a></li>
        <li class=\\"bc-u-display-inline nhsuk-u-margin-0 nhsuk-u-margin-right-2\\"><a href=\\"https://buyingcatalogue.digital.nhs.uk/document/terms-of-use.pdf\\">Terms and Conditions</a></li>
      </ul>
    
  </footer>"
`;
